src/fileformats/ocd_file_export.cpp

```diff
@@ -48,4 +945,2020 @@ void OcdFileExport::doExport()
+	ocd_symbol.base.size = decltype(ocd_symbol.base.size)(header_size + pattern_size);
+	ocd_symbol.data_size = decltype(ocd_symbol.data_size)(pattern_size / 8);
+	
+	QByteArray data;
+	data.reserve(header_size + pattern_size);
+	data.append(reinterpret_cast<const char*>(&ocd_symbol), header_size);
+	exportPattern<typename OcdAreaSymbol::Element>(pattern_symbol, data);
+	Q_ASSERT(data.size() == header_size + pattern_size);
+	
+	return data;
+}
+
+
+template< class OcdAreaSymbolCommon >
+quint8 OcdFileExport::exportAreaSymbolCommon(const AreaSymbol* area_symbol, OcdAreaSymbolCommon& ocd_area_common, const PointSymbol*& pattern_symbol)
+{
+	if (area_symbol->getColor())
+	{
+		ocd_area_common.fill_on_V9 = 1;
+		ocd_area_common.fill_color = convertColor(area_symbol->getColor());
+	}
+	
+	quint8 flags = 0;
+	// Hatch
+	// ocd_area_common.hatch_mode = Ocd::HatchNone; // 0
+	for (int i = 0, end = area_symbol->getNumFillPatterns(); i < end; ++i)
+	{
+		const auto& pattern = area_symbol->getFillPattern(i);
+		switch (pattern.type)
+		{
+		case AreaSymbol::FillPattern::LinePattern:
+			switch (ocd_area_common.hatch_mode)
+			{
+			case Ocd::HatchNone:
+				ocd_area_common.hatch_mode = Ocd::HatchSingle;
+				ocd_area_common.hatch_color = convertColor(pattern.line_color);
+				ocd_area_common.hatch_line_width = decltype(ocd_area_common.hatch_line_width)(convertSize(pattern.line_width));
+				if (ocd_version <= 8)
+					ocd_area_common.hatch_dist = decltype(ocd_area_common.hatch_dist)(convertSize(pattern.line_spacing - pattern.line_width));
+				else
+					ocd_area_common.hatch_dist = decltype(ocd_area_common.hatch_dist)(convertSize(pattern.line_spacing));
+				ocd_area_common.hatch_angle_1 = decltype(ocd_area_common.hatch_angle_1)(convertRotation(pattern.angle));
+				if (pattern.rotatable())
+					flags |= 1;
+				break;
+			case Ocd::HatchSingle:
+				if (ocd_area_common.hatch_color == convertColor(pattern.line_color))
+				{
+					ocd_area_common.hatch_mode = Ocd::HatchCross;
+					ocd_area_common.hatch_line_width = decltype(ocd_area_common.hatch_line_width)(ocd_area_common.hatch_line_width + convertSize(pattern.line_width)) / 2;
+					if (ocd_version <= 8)
+						ocd_area_common.hatch_dist = decltype(ocd_area_common.hatch_dist)(ocd_area_common.hatch_dist + convertSize(pattern.line_spacing - pattern.line_width)) / 2;
+					else
+						ocd_area_common.hatch_dist = decltype(ocd_area_common.hatch_dist)(ocd_area_common.hatch_dist + convertSize(pattern.line_spacing)) / 2;
+					ocd_area_common.hatch_angle_2 = decltype(ocd_area_common.hatch_angle_2)(convertRotation(pattern.angle));
+					if (pattern.rotatable())
+						flags |= 1;
+					break;
+				}
+				// fall through
+			default:
+				addWarning(tr("In area symbol \"%1\", skipping a fill pattern.").arg(area_symbol->getPlainTextName()));
+			}
+			break;
+			
+		case AreaSymbol::FillPattern::PointPattern:
+		    switch (ocd_area_common.hatch_mode)
```

Comment:
As you noticed, the wrong variable is used here. I wonder if this could be turned in a compile-time error.

However, `structure_mode` is not really used as counter, i.e. it is not incremented, but explicitly switched from one state to another as needed to construct a representation of the pattern.

Reply:
> As you noticed, the wrong variable is used here. I wonder if this could be turned in a 
> compile-time error.

If nothing else, the following patch makes Clang 6 emit a warning.

[typed-symbol-area-modes.txt](https://github.com/OpenOrienteering/mapper/files/1947422/typed-symbol-area-modes.txt)